app/src/swap.c

```diff
@@ -0,0 +1,290 @@
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+********************************************************************************/
+#include "swap.h"
+#include "crypto.h"
+#include "bignum.h"
+#include "zxformat.h"
+#include "substrate_dispatch.h"
+
+static zxerr_t extractHDPath(uint8_t* params, uint8_t paramsSize);
+static zxerr_t readU32BE(uint8_t* input, uint32_t *output);
+static zxerr_t readU64BE(uint8_t* input, uint64_t *output);
+static zxerr_t bytesAmountToStringBalance(uint8_t *amount, uint8_t amount_len, char *out, uint8_t out_len);
+
+static void handle_check_address(check_address_parameters_t* params);
+static void handle_get_printable_amount(get_printable_amount_parameters_t* params);
+
+static bool copy_transaction_parameters(create_transaction_parameters_t* sign_transaction_params);
+
+swap_globals_t G_swap_state;
+
+void swap_handle_command(libargs_s *args) {
+    switch (args->command) {
+        case check_address:
+            handle_check_address(args->check_address);
+            G_swap_state.called_from_swap = 0;
+            break;
+
+        case sign_transaction:
+            if (copy_transaction_parameters(args->create_transaction)) {
+                G_swap_state.called_from_swap = 1;
+            }
+            break;
+
+        case get_printable_amount:
+            handle_get_printable_amount(args->get_printable_amount);
+            G_swap_state.called_from_swap = 0;
+            break;
+
+        case run_application:
+        default:
+            break;
+    }
+}
+
+void handle_check_address(check_address_parameters_t* params) {
+    if (params == NULL) {
+        return;
+    }
+    params->result = 0;
+    uint8_t buffer[100] = {0};
+    uint16_t replyLen = 0;
+
+    // address parameters have the following structure
+    // address kind (1 byte) | path length (1 byte) | bip44 path (4 * pathLength bytes)
+    const key_kind_e add_kind = (key_kind_e) *params->address_parameters;
+    zxerr_t err = extractHDPath((uint8_t*) params->address_parameters + 2, (uint8_t) params->address_parameters_length - 2);
+    if (params->address_to_check == 0 || err != zxerr_ok) {
+        return;
+    }
+
+    err = crypto_fillAddress(add_kind, buffer, sizeof(buffer), &replyLen);
+
+    if (err != zxerr_ok || replyLen <= PK_LEN_25519) {
+        MEMZERO(buffer, sizeof(buffer));
+        return;
+    }
+
+    const uint8_t *address = buffer + PK_LEN_25519;
+    const uint8_t addressLen = replyLen - PK_LEN_25519;
+    const uint8_t addressToCheckLen = strlen(params->address_to_check);
```

Comment:
It would be better to use strnlen instead of strlen but we don't have the size of the address_to_check field